### PR TITLE
Adds `@JsonKey` annotation.

### PR DESCRIPTION
**Pull request by Anusien**
This annotation will be used like `@JsonValue` but only when
serializing Map keys.

See jackson-databind/#2871

### CODE REVIEW

File: src/main/java/com/fasterxml/jackson/annotation/JsonKey.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+ *<p>
+ * A typical usage is that of annotating <code>toString()</code>
+ * method so that returned String value is used as the JSON serialization;
+ * and if deserialization is needed, there is matching constructor
+ * or factory method annotated with {@link JsonCreator} annotation.
+ *<p>
+ * Boolean argument is only used so that sub-classes can "disable"
+ * annotation if necessary.
+ *<p>
+ * NOTE: when use for Java <code>enum</code>s, one additional feature is
```
Comment (cowtowncoder) [inline src/main/java/com/fasterxml/jackson/annotation/JsonKey.java:34]: Probably needs to be removed for now

File: src/main/java/com/fasterxml/jackson/annotation/JsonKey.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+ * that value returned by annotated method is also considered to be the
+ * value to deserialize from, not just JSON String to serialize as.
+ * This is possible since set of Enum values is constant and it is possible
+ * to define mapping, but can not be done in general for POJO types; as such,
+ * this is not used for POJO deserialization.
+ *
+ * @see JsonCreator
```
Comment (cowtowncoder) [inline src/main/java/com/fasterxml/jackson/annotation/JsonKey.java:41]: Not sure `@JsonCreator` handling would be yet supported either.

**Comment by cowtowncoder** [comment]
Aside from javadoc comments (which I can tweak too), looks good.

I don't think I have CLA yet from you? If so, it's here:

https://github.com/FasterXML/jackson/blob/master/contributor-agreement.pdf

and usual way is to print, fill & sign, scan/photo, email to `info` at fasterxml dot com.
Only needs to be done once, and after I get it works for all Jackson contributions. The only piece of process needed here fortunately. :)

**Comment by Anusien** [comment]
Indeed's signed a corporate CLA, should be covered.

**Pull request merged**